### PR TITLE
add support for skipping tasks - fixes #15

### LINKED ISSUE #15: Skippable Tasks
Allow tasks to be labelled as "skipped" instead of completed, with an optional message explaining why.

See https://github.com/sindresorhus/np/pull/64#issuecomment-233775749

The explanation message should probably persist instead of being swallowed by `log-update`.

### PR DESCRIPTION
**Pull request by unkillbob**
skipping

Second attempt at skip task support (after https://github.com/SamVerschueren/listr/pull/17), this time with an explicit `skip()` function as per https://github.com/SamVerschueren/listr/pull/17#issuecomment-234362295.

> - if skip returns false or a Promise for false, continue execution.
> - if skip returns true or a Promise for true, skip with a generic message.
> - if skip returns a string or a Promise for a string, skip with a custom message.
> - if skip throws or returns a rejected promise, fail the entire build

An example of how this could be used in https://github.com/sindresorhus/np:

```js
{
	title: 'Publishing package',
	skip: () => {
		if (readPkgUp.sync().pkg.private) {
			return 'Private package: not publishing to npm.';
		}
	},
	task: () => exec('npm', ['publish'].concat(opts.tag ? ['--tag', opts.tag] : []))
}
```

screen shot 2016-07-24 at 10 28 31 pm

The tests could do with splitting up (probably to align nicely with the bulleted list of behaviour outlined above) and the readme needs updating but would be good to make sure everyone's happy with this solution before I get those done.

### CODE REVIEW

File: lib/renderer/string-renderer.js [hunk 1]
```diff
@@ -1,4 +1,7 @@
 'use strict';
+
```
Comment (SamVerschueren) [inline lib/renderer/string-renderer.js]: remove empty line

File: lib/task.js [hunk 2]
```diff
@@ -46,11 +51,16 @@ class Task {
 			throw new TypeError(`Expected property \`task\` of type \`function\` but got \`${typeof task.task}\``);
 		}
 
+		if (typeof task.skip !== 'undefined' && typeof task.skip !== 'function') {
```
Comment (SamVerschueren) [inline lib/task.js]:
```js
if (task.skip && typeof task.skip !== 'function') {
```

File: lib/task.js [hunk 3]
```diff
@@ -80,6 +91,16 @@ class Task {
 		return Promise.resolve()
 			.then(() => {
 				this.state = state.PENDING;
+				return this.skip();
+			})
+			.then(skip => {
```
Comment (SamVerschueren) [inline lib/task.js]: `skip` -> `skipped`

File: lib/task.js [hunk 4]
```diff
@@ -80,6 +91,16 @@ class Task {
...
+				if (skip) {
+					const reason = typeof skip === 'string' ? skip : `Skipping ${this.title}.`;
```
Comment (SamVerschueren) [inline lib/task.js]: I don't think we should print a reason if no reason was provided by the users themselves.

File: lib/task.js [hunk 5]
```diff
@@ -111,7 +132,9 @@ class Task {
 				return result;
 			})
 			.then(() => {
-				this.state = state.COMPLETED;
+				if (this.state !== state.SKIPPED) {
```
Comment (SamVerschueren) [inline lib/task.js]:
```js
if (this.state === state.PENDING) {
```

In case we add more states in the future.

File: test.js [hunk 6]
```diff
@@ -85,3 +85,45 @@ test('add tasks', t => {
 
 	t.is(list._tasks.length, 4);
 });
+
+test('skip tasks', t => {
+	t.plan(5);
```
Comment (SamVerschueren) [inline test.js]: remove `t.plan`

File: test.js [hunk 7]
```diff
@@ -85,3 +85,45 @@ test('add tasks', t => {
 
 	t.is(list._tasks.length, 4);
 });
+
+test('skip tasks', t => {
```
Comment (SamVerschueren) [inline test.js]: `test('skip tasks', async t => {`

File: test.js [hunk 8]
```diff
@@ -85,3 +85,45 @@ test('add tasks', t => {
...
+			task: () => t.fail('Skipping should not execute the task')
+		},
+		{
+			title: 'Task 4',
+			skip: () => Promise.resolve(false),
+			task: () => t.pass() || '4'
+		}
+	]);
+
+	return list.run();
```
Comment (SamVerschueren) [inline test.js]:
```js
await list.run();
```

File: test.js [hunk 9]
```diff
@@ -85,3 +85,45 @@ test('add tasks', t => {
...
 });
+
+test('skip tasks', t => {
+	t.plan(5);
+
+	const list = new Listr([
+		{
+			title: 'Task 1',
+			task: () => new Listr([
+				{title: 'Task 1.1', task: () => t.pass() || '1.1'},
```
Comment (SamVerschueren) [inline test.js]: Why did you add the `|| '1.1'`?

**Comment by SamVerschueren** [comment]
Super excited about this feature!

**Comment by SamVerschueren** [comment]
Should we add `chalk.dim` to the `[Skipped]` part at the end of the line?

File: test.js (hunk 9 above)
Comment (unkillbob) [inline test.js]: I had the `t.plan(5)` and the `t.pass()` to ensure that all the non-skipped tasks still ran. 

`t.pass()` doesn't return a value and currently the implementation of Listr requires the `task()` function return a result so that's why I returned *something* from each of the `task` functions.

But as mentioned in the PR description I'm keen to give the tests some love, this was just to prove the concept really.

Comment (SamVerschueren) [inline test.js]: > I had the t.plan(5) and the t.pass() to ensure that all the non-skipped tasks still ran.

That's a valid point :). But can you change it to use `async`/`await` then? You can keep the `t.plan(5)` as that indeed makes sense.

> currently the implementation of Listr requires the task() function return a result

That indeed was the case but I fixed that yesterday as you can see [here](https://github.com/SamVerschueren/listr/blob/master/lib/task.js#L17-L19). It doesn't fail anymore if `result` is `undefined. So I believe it should work without the `|| '1.1';` stuff. If not, let me know because that should be fixed then.

> But as mentioned in the PR description I'm keen to give the tests some love, this was just to prove the concept really.

Very much appreciated :)!

File: lib/task.js [hunk 10]
```diff
@@ -80,6 +91,17 @@ class Task {
 		return Promise.resolve()
 			.then(() => {
 				this.state = state.PENDING;
+				return this.skip();
+			})
+			.then(skipped => {
+				if (skipped) {
+					this.state = state.SKIPPED;
+					if (typeof skipped === 'string') {
+						this._lastLine = stripAnsi(skipped.trim().split('\n').pop().trim());
```
Comment (SamVerschueren) [inline lib/task.js]: Could you remove the `stripAnsi` call at this line and at line 125. I think it would make more sense to call it where the actual logging is done at line 83.

**Comment by unkillbob** [comment]
@SamVerschueren I think this is ready for another look.

1. Tidied up and added more tests
2. Added skip docs to readme
3. Dimmed the `[skipped]` annotation in the title

File: readme.md [hunk 11]
```diff
@@ -135,6 +135,37 @@ const tasks = new Listr([
...
+```js
+const tasks = new Listr([
+	{
+		title: 'Task 1',
+		task: () => Promise.resolve('Foo')
+	},
+	{
+		title: 'Can be skipped',
+		skip: () => {
+			if (someCondition) {
```
Comment (SamVerschueren) [inline readme.md]:
```js
if (Math.random() > 0.5) {
    return 'Reason for skipping';
}
```

File: readme.md [hunk 12]
```diff
@@ -135,6 +135,37 @@ const tasks = new Listr([
...
+				return 'Reason for skipping';
+			}
+			return false;
```
Comment (SamVerschueren) [inline readme.md]: Remove the `return false`

**Comment by SamVerschueren** [comment]
We're almost there I guess. Can you put the task objects on different lines in the tests?

```js
{title: 'title', skip: () => false, task: () => { // task }}
```

=>

```js
{
    title: 'title',
    skip: () => false,
    task: () => {
        // task
    }
}
```

Much easier to reason about the logic.

@sindresorhus @jamestalmage Care to take a look at it?

**Comment by sindresorhus** [comment]
LGTM

File: readme.md [hunk 13]
```diff
@@ -157,6 +187,12 @@ Type: `Function`
 
 Task function.
 
+##### skip
+
+Type: `Function`
+
+Skip function.
```
Comment (SamVerschueren) [inline readme.md]: Should we show the different possibilities here? Otherwise it will be difficult for users to know what it does in the different scenario's.

> - if skip returns false or a Promise for false, continue execution.
> - if skip returns true or a Promise for true, skip with a generic message.
> - if skip returns a string or a Promise for a string, skip with a custom message.
> - if skip throws or returns a rejected promise, fail the entire build

After this, it's good to merge I guess 🎉  👍  ❤️

Comment (unkillbob) [inline readme.md]: I cover it in more detail on line 142, following the same pattern for the `task` function (which had more detail in the prose docs, and less detail in the API docs). Happy to change it, up to you 👍

Comment (SamVerschueren) [inline readme.md]: Oh yes. Maybe we should link to it then. 

> Read more about [skipping tasks]()

**Comment by SamVerschueren** [comment]
Super great work @unkillbob ! Very much appreciate the effort 🍻
